python/lsst/ip/diffim/dipoleFitTask.py

```diff
@@ -0,0 +1,921 @@
+
+        # Third, is it a good fit (chi2dof < 1)?
+        # Use scipy's chi2 cumulative distrib to estimate significance
+        # This doesn't really work since I don't trust the values in the variance plane (which
+        #   affects the least-sq weights, which affects the resulting chi2).
+        # But I'm going to keep this here for future use.
+        if False:
+            from scipy.stats import chi2
+            ndof = result.psfFitChi2 / measRecord[self.chi2dofKey]
+            significance = chi2.cdf(result.psfFitChi2, ndof)
+            passesChi2 = significance < self.config.maxChi2DoF
+            allPass = allPass and passesChi2
+
+        if allPass:  # Note cannot pass `allPass` into the `measRecord.set()` call below...?
+            measRecord.set(self.classificationFlagKey, True)
+        else:
+            measRecord.set(self.classificationFlagKey, False)
+
+    # TBD: need to catch more exceptions, set correct flags.
+    def fail(self, measRecord, error=None):
+        measRecord.set(self.flagKey, True)
+        if error is not None:
+            if error.getFlagBit() == self.FAILURE_EDGE:
+                measRecord.set(self.edgeFlagKey, True)
+            if error.getFlagBit() == self.FAILURE_FIT:
+                measRecord.set(self.flagKey, True)
+            if error.getFlagBit() == self.FAILURE_NOT_DIPOLE:
+                measRecord.set(self.flagKey, True)
+
+
+class DipoleFitAlgorithm(object):
+    """Lightweight class containing methods for fitting dipoles in
+    diffims, used by DipoleFitPlugin.  This code is documented in
+    DMTN-007.
+
+    Below is a (somewhat incomplete) list of improvements
+    that would be worth investigating, given the time:
+
+    1. evaluate necessity for separate parameters for pos- and neg- images
+    2. only fit background OUTSIDE footprint (DONE) and dipole params INSIDE footprint (NOT DONE)?
+    3. correct normalization of least-squares weights based on variance planes
+    4. account for PSFs that vary across the exposures (should be happening by default?)
+    5. correctly account for NA/masks  (i.e., ignore!)
+    6. better exception handling in the plugin
+    7. better classification of dipoles (e.g. by comparing chi2 fit vs. monopole?)
+    8. (DONE) Initial fast estimate of background gradient(s) params -- perhaps using numpy.lstsq
+    9. (NOT NEEDED - see (1)) Initial fast test whether a background gradient needs to be fit
+    10. (DONE) better initial estimate for flux when there's a strong gradient
+    11. (DONE) requires a new package `lmfit` -- investiate others? (astropy/scipy/iminuit?)
+    """
+
+    # This is just a private version number to sync with the ipython notebooks that I have been
+    # using for algorithm development.
+    _private_version_ = '0.0.4'
+
+    # Create a namedtuple to hold all of the relevant output from the lmfit results
+    resultsOutput = namedtuple('resultsOutput',
```

Comment:
Please use pipe_base Struct, instead, so users cannot mess up by assuming an order. An alternative if you want to pre-define the fields is to use a small custom class.